src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -486,53 +479,191 @@ private static void SetLegacyResourceNames(BeforeActionStruct typedArg, Span spa
+                if (parentSpan.Tags is AspNetCoreEndpointTags parentTags)
                 {
-                    absolutePath = request.PathBase.Value + absolutePath;
+                    parentTags.AspNetCoreRoute = aspNetRoute;
                 }
 
-                string resourceUrl = UriHelpers.GetCleanUriPath(absolutePath)
-                                               .ToLowerInvariant();
+                parentSpan.ResourceName = span.ResourceName;
+            }
 
-                string resourceName = $"{httpMethod} {resourceUrl}";
+            return span;
+        }
 
-                SpanContext propagatedContext = ExtractPropagatedContext(request);
-                var tagsFromHeaders = ExtractHeaderTags(request, tracer);
+        private Span StartCoreSpan(Tracer tracer, HttpContext httpContext, HttpRequest request)
+        {
+            string host = request.Host.Value;
+            string httpMethod = request.Method?.ToUpperInvariant() ?? "UNKNOWN";
+            string url = request.GetUrl();
 
-                var tags = tracer.Settings.RouteTemplateResourceNamesEnabled ? new AspNetCoreEndpointTags() : new AspNetCoreTags();
-                var scope = tracer.StartActiveWithTags(HttpRequestInOperationName, propagatedContext, tags: tags);
+            if (tracer.Settings.RouteTemplateResourceNamesEnabled)
+            {
+                httpContext.Features.Set(new RequestTrackingFeature
+                {
+                    HttpMethod = httpMethod,
+                    OriginalUrl = url,
+                });
+            }
 
-                scope.Span.DecorateWebServerSpan(resourceName, httpMethod, host, url, tags, tagsFromHeaders);
+            string absolutePath = request.Path.Value;
 
-                tags.SetAnalyticsSampleRate(IntegrationId, tracer.Settings, enabledWithGlobalSetting: true);
+            if (request.PathBase.HasValue)
+            {
+                absolutePath = request.PathBase.Value + absolutePath;
+            }
+
+            string resourceUrl = UriHelpers.GetCleanUriPath(absolutePath)
+                                           .ToLowerInvariant();
+
+            string resourceName = $"{httpMethod} {resourceUrl}";
+
+            SpanContext propagatedContext = ExtractPropagatedContext(request);
+            var tagsFromHeaders = ExtractHeaderTags(request, tracer);
+
+            var tags = tracer.Settings.RouteTemplateResourceNamesEnabled ? new AspNetCoreEndpointTags() : new AspNetCoreTags();
+            var scope = tracer.StartActiveWithTags(HttpRequestInOperationName, propagatedContext, tags: tags);
+
+            scope.Span.DecorateWebServerSpan(resourceName, httpMethod, host, url, tags, tagsFromHeaders);
+
+            tags.SetAnalyticsSampleRate(IntegrationId, tracer.Settings, enabledWithGlobalSetting: true);
+            Log.Warning($"Created span {scope.Span.SpanId} with {resourceName}, {scope.Span.ResourceName}");
+            return scope.Span;
+        }
```

Comment:
This should be removed

Reply:
Not sure what you mean, this function is need to raise the span and the library won't compile without it.

Reply:
Oops, sorry @robertpi,  it wasn't supposed to highlight the whole method (using vs integration to review)!

This: needs to be removed, and could be a big part of the performance regression:
```
            Log.Warning($"Created span {scope.Span.SpanId} with {resourceName}, {scope.Span.ResourceName}");
```

Reply:
Guess, you mean the warning? Removed in c2f204a6